src/realm/sync/network/default_socket.cpp

```diff
@@ -371,10 +378,177 @@ void DefaultWebSocketImpl::initiate_websocket_handshake()
 }
 } // namespace
 
+///
+/// DefaultSocketProvider - default socket provider implementation
+///
+
+DefaultSocketProvider::DefaultSocketProvider(const std::shared_ptr<util::Logger>& logger,
+                                             const std::string user_agent)
+    : m_logger_ptr{logger}
+    , m_service{}
+    , m_random{}
+    , m_user_agent{user_agent}
+    , m_mutex{}
+    , m_state{State::Stopped}
+    , m_state_cv{}
+    , m_thread{}
+{
+    REALM_ASSERT(m_logger_ptr);                     // Make sure the logger is valid
+    util::seed_prng_nondeterministically(m_random); // Throws
+    start_keep_running_timer();                     // TODO: Update service so this timer is not needed
+    start();
+}
+
+DefaultSocketProvider::~DefaultSocketProvider()
+{
+    // Wait for the thread to stop
+    stop(true);
+    if (m_keep_running_timer)
+        m_keep_running_timer->cancel();
+    if (m_thread.joinable()) {
+        m_thread.join();
+    }
+    // Shutting down - no need to lock mutex before check
+    REALM_ASSERT(m_state == State::Stopped);
+}
+
+void DefaultSocketProvider::start()
+{
+    std::unique_lock<std::mutex> lock(m_mutex);
+    // Has the thread already been started or is running
+    if (m_state == State::Starting || m_state == State::Started || m_state == State::Running)
+        return; // early return
+
+    // If the thread has been previously run, make sure it has been joined first
+    if (m_state == State::Stopping || m_state == State::Stopped) {
+        if (m_thread.joinable()) {
+            lock.unlock();
+            m_thread.join();
+            lock.lock();
+        }
+    }
+
+    m_logger_ptr->trace("Default event loop: start()");
+    REALM_ASSERT(m_state == State::Stopped);
+    do_state_update(State::Starting);
+    m_thread = std::thread{make_event_loop()};
+    lock.unlock();
+    // Wait for the thread to start before continuing
+    state_wait_for(State::Started);
+}
+
+std::function<void()> DefaultSocketProvider::make_event_loop()
+{
+    return [this]() {
+        m_logger_ptr->trace("Default event loop: thread running");
+        std::unique_lock<std::mutex> lock(m_mutex);
+        // The thread has started - don't return early until after did_create_thread()
+        if (m_state == State::Starting) {
+            do_state_update(State::Started);
+        }
+        lock.unlock();
+
+        auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
+            m_logger_ptr->trace("Default event loop: thread exiting");
+            if (g_binding_callback_thread_observer)
+                g_binding_callback_thread_observer->will_destroy_thread();
+
+            {
+                std::lock_guard<std::mutex> lock(m_mutex);
+                if (m_state != State::Stopping)
+                    return; // early return
```

Comment:
When could this happen?

Reply:
It shouldn't ever happen - I updated it to a `REALM_ASSERT()`